aiagents4pharma/talk2knowledgegraphs/utils/extractions/multimodal_pcst.py

```diff
@@ -0,0 +1,292 @@
+"""
+Exctraction of multimodal subgraph using Prize-Collecting Steiner Tree (PCST) algorithm.
+"""
+
+from typing import Tuple, NamedTuple
+import numpy as np
+import pandas as pd
+import torch
+import pcst_fast
+from torch_geometric.data.data import Data
+
+class MultimodalPCSTPruning(NamedTuple):
+    """
+    Prize-Collecting Steiner Tree (PCST) pruning algorithm implementation inspired by G-Retriever
+    (He et al., 'G-Retriever: Retrieval-Augmented Generation for Textual Graph Understanding and
+    Question Answering', NeurIPS 2024) paper.
+    https://arxiv.org/abs/2402.07630
+    https://github.com/XiaoxinHe/G-Retriever/blob/main/src/dataset/utils/retrieval.py
+
+    Args:
+        topk: The number of top nodes to consider.
+        topk_e: The number of top edges to consider.
+        cost_e: The cost of the edges.
+        c_const: The constant value for the cost of the edges computation.
+        root: The root node of the subgraph, -1 for unrooted.
+        num_clusters: The number of clusters.
+        pruning: The pruning strategy to use.
+        verbosity_level: The verbosity level.
+    """
+    topk: int = 3
+    topk_e: int = 3
+    cost_e: float = 0.5
+    c_const: float = 0.01
+    root: int = -1
+    num_clusters: int = 1
+    pruning: str = "gw"
+    verbosity_level: int = 0
+    use_description: bool = False
+
+    def _compute_node_prizes(self,
+                             graph: Data,
+                             query_emb: torch.Tensor,
+                             modality: str) :
+        """
+        Compute the node prizes based on the cosine similarity between the query and nodes.
+
+        Args:
+            graph: The knowledge graph in PyTorch Geometric Data format.
+            query_emb: The query embedding in PyTorch Tensor format. This can be an embedding of
+                a prompt, sequence, or any other feature to be used for the subgraph extraction.
+            modality: The modality to use for the subgraph extraction based on the node type.
+
+        Returns:
+            The prizes of the nodes.
+        """
+        # Convert PyG graph to a DataFrame
+        graph_df = pd.DataFrame({
+            "node_type": graph.node_type,
+            "desc_x": [x.tolist() for x in graph.desc_x],
+            "x": [list(x) for x in graph.x],
+            "score": [0.0 for _ in range(len(graph.node_id))],
+        })
+
+        # Calculate cosine similarity for text features and update the score
+        if self.use_description:

```

Comment:
Please correct me if I am wrong, but this looks like something that could be off loaded to a vector database (or just using CuVS if you want to store the index in memory) as the Prize is just the similarity score?